src/usuba_ES.ml

```diff
@@ -40,22 +39,48 @@ module Uexpr = struct
     let compare h1 h2 = Int.compare h1.Hash_union.tag h2.Hash_union.tag
   end
 
-  module HSet = Set.Make (HCons)
+  module HSet = struct
+    include Set.Make (HCons)
+
+    let pp ?(pp_sep = fun ppf () -> Format.fprintf ppf "; ") ?(left = "{")
+        ?(right = "}") () ppf t =
+      Hash_union.Printers.(
+        pp_list ~pp_sep ~left ~right Hash_union.(pp_hash_consed Hash_expr.pp))
+        ppf (elements t)
+  end
 
   type t = { union : Hash_union.PUnion.t; exprs : HSet.t }
 
+  let pp ppf t =
+    Format.fprintf ppf "@[<v 1>{union: %a;@ exprs: %a}@]" Hash_union.PUnion.pp
+      t.union
+      HSet.(pp ())
+      t.exprs
+
   let create n = { union = Hash_union.PUnion.create n; exprs = HSet.empty }
   let find e u = Hash_union.PUnion.find e.union u.Hash_union.tag
 
-  let union h x y =
+  let union { union; exprs } x y =
+    let open Hash_union in
+    let tag_x = x.tag in
+    let tag_y = y.tag in
+    let mtag = max tag_x tag_y in
+    (* Union structure stores the tags *)
+    (* If we add a tag that is greater than the size of the union structure *)
+    (* all the previous equivalence classes are the same but the new one needs to be *)
+    (* added in a bigger union. Just resize it (copying the previous union in it) *)
+    (* and proceeds *)
+    let union =
+      if mtag > PUnion.length union then PUnion.expand union mtag else union
```

Comment:
Very expensive way to resize an array. 

Also, this is too loose a way to keep track of the invariants between (successive) tag members of hashconsed objects, their insertion in `.exprs` and the equivalence classes in `.union`.